### ISSUE TITLE
ethereumjs-vm 2.0.1 compatibility.

### ISSUE THREAD

**Issue by simondlr** [ConsenSys-archive/ganache-cli-archive#197]
I upgraded to use the latest version of the vm (2.0.1). However, with it came issues wrt state trie lookups. Getting an issue wrt funds not being available.

```Error: sender doesn't have enough funds to send tx. The upfront cost is: 471238800000000000 and the senders account only has: 0```

Thanks to @mhhf (Denis), he pointed me in the right direction. It's a simple fix since the interface of the VM changed.

At blockchain_double.js, change the following part from:

```
  this.vm = options.vm || new VM(this.stateTrie, this, {
    enableHomestead: true,
    activatePrecompiles: true
  });
```

to

```
  this.vm = options.vm || new VM({
    state: this.stateTrie,
    enableHomestead: true,
    activatePrecompiles: true
  });
```

This correctly passes the correct state trie.

I can submit a PR, but it seems there are more extensive code/writing around blockchain state in testrpc, so not sure what takes precedence atm. Let me know.

**Comment by simondlr** [comment]
I added a PR anyway. :)

**Comment by tcoulter** [comment]
Most tests pass with your PR. I need to research the three failing tests; they all revolve around a specific blockhash that's generate. It could just be those tests were poorly written.

**Comment by jmdfm** [comment]
Here are the 3 failing tests in question: 

```
150 passing (13s)
  3 failing

  1) Forking should represent the block number correctly in the Oracle contract (oracle.blockhash0), providing forked block hash and number:

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x1e27de32f8daf6d4adc903b68282e866dd6bdf3409d4b4976885f4cfa5dd399e'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x1e27de32f8daf6d4adc903b68282e866dd6bdf3409d4b4976885f4cfa5dd399e

      at test/forking.js:373:18

  2) Provider: contract scenario should represent the block number correctly in the Oracle contract (oracle.blockhash0):

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x6dd2f0ee6e56750dfb0bc20f9dab7a1833def6c5dd17479d495b3fe95d84ac2c'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x6dd2f0ee6e56750dfb0bc20f9dab7a1833def6c5dd17479d495b3fe95d84ac2c

      at test/requests.js:402:20

  3) Server: contract scenario should represent the block number correctly in the Oracle contract (oracle.blockhash0):

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x6481e0630e8db611eb87f9c51efea96f9a705b67843c9cb52cf47d89741cf82d'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x6481e0630e8db611eb87f9c51efea96f9a705b67843c9cb52cf47d89741cf82d

      at test/requests.js:402:20
      at endReadableNT (_stream_readable.js:974:12)
      at _combinedTickCallback (internal/process/next_tick.js:74:11)
      at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by jmdfm** [comment]
What's notable about these failures is that all 2 tests expect a block hash of `0x044852b2a670ade5`, but the each test gets a different block hash.

**Comment by jmdfm** [comment]
Subsequent runs of the tests in question show that a different block hash is returned every time:

```
13:24 [~/...nsensys/testrpc simondlr-master *%] node_modules/mocha/bin/mocha test/requests.js


  Server:
    contract scenario
      1) should represent the block number correctly in the Oracle contract (oracle.blockhash0)


  0 passing (477ms)
  1 failing

  1) Server: contract scenario should represent the block number correctly in the Oracle contract (oracle.blockhash0):

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x138e4753165e36ae1902a7e805e775f8847dc4819a52fb3b8c5a5f7bfbf6570c'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x138e4753165e36ae1902a7e805e775f8847dc4819a52fb3b8c5a5f7bfbf6570c

      at test/requests.js:402:20
      at endReadableNT (_stream_readable.js:974:12)
      at _combinedTickCallback (internal/process/next_tick.js:74:11)
      at process._tickCallback (internal/process/next_tick.js:98:9)



13:24 [~/...nsensys/testrpc simondlr-master *%] node_modules/mocha/bin/mocha test/requests.js


  Server:
    contract scenario
      1) should represent the block number correctly in the Oracle contract (oracle.blockhash0)


  0 passing (482ms)
  1 failing

  1) Server: contract scenario should represent the block number correctly in the Oracle contract (oracle.blockhash0):

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x06500f0e92c3e30c7f082b28ab5be872de3bd0a80d8859a27d98e44c569daa47'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x06500f0e92c3e30c7f082b28ab5be872de3bd0a80d8859a27d98e44c569daa47

      at test/requests.js:402:20
      at endReadableNT (_stream_readable.js:974:12)
      at _combinedTickCallback (internal/process/next_tick.js:74:11)
      at process._tickCallback (internal/process/next_tick.js:98:9)



13:24 [~/...nsensys/testrpc simondlr-master *%] node_modules/mocha/bin/mocha test/requests.js


  Server:
    contract scenario
      1) should represent the block number correctly in the Oracle contract (oracle.blockhash0)


  0 passing (482ms)
  1 failing

  1) Server: contract scenario should represent the block number correctly in the Oracle contract (oracle.blockhash0):

      Uncaught AssertionError: '0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d' == '0x37f1e64da5e6c2f988c8f2ec9071da51fc2a30ef1e71ebc1d60abac4001736e2'
      + expected - actual

      -0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d
      +0x37f1e64da5e6c2f988c8f2ec9071da51fc2a30ef1e71ebc1d60abac4001736e2

      at test/requests.js:402:20
      at endReadableNT (_stream_readable.js:974:12)
      at _combinedTickCallback (internal/process/next_tick.js:74:11)
      at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by jmdfm** [comment]
So the oracle contracts block hash is always `0x044852b2a670ade5407e78fb2863c51de9fcb96542a07186fe3aeda6bb8a116d` but the Block 0 block hash (the results of calling `web3.eth.getBlock(0, etc)` is changing with every invocation

**Comment by jmdfm** [comment]
Fixed in PR #205
